File: src/lib/scala/Slate.Common/src/main/scala/slate/common/ApiKey.scala
/**
<slate_header>
  author: Kishore Reddy
  url: https://github.com/kishorereddy/scala-slate
  copyright: 2015 Kishore Reddy
  license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  desc: a scala micro-framework
  usage: Please refer to license on github for more info.
</slate_header>
  */

package slate.common

/**
 * Represents a key for accessing an API securely.
 * @param name : "admin"
 * @param key  : "123456789123456789"
 * @param rolesLookup : "admin" -> true, "dev" -> true
 */
case class ApiKey(
                   name:String,
                   key :String,
                   rolesLookup:Map[String,String]
                 )
{
}


object ApiKey {

  def apply(name:String, key:String, roles:String): ApiKey = {
    ApiKey(name, key, Option(roles).fold(Map[String,String]())( r => Strings.splitToMap(r, ',', true)))
  }
}

File: src/common/common/src/main/kotlin/kiit/common/info/ApiKey.kt
/**
 *  <kiit_header>
 * url: www.kiit.dev
 * git: www.github.com/slatekit/kiit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 *
 * 
 *  </kiit_header>
 */

package kiit.common.info

/**
 * Represents a key for accessing an API securely.
 * @param name : "admin"
 * @param key : "123456789123456789"
 * @param rolesLookup : "admin" -> true, "dev" -> true
 */
data class ApiKey(
    @JvmField val name: String,
    @JvmField val key: String,
    @JvmField val roles: String,
    @JvmField val rolesLookup: Map<String, String>
) {

    constructor(name: String, key: String, roles: String) : this(name, key, roles,
            build(roles)
    )

    companion object {

        @JvmStatic
        fun build(roles: String?): Map<String, String> {
            return roles?.splitToSequence(',')?.associateBy({ it }, { it }) ?: mapOf()
        }

        @JvmStatic
        fun build(name: String, key: String, roles: String?): ApiKey {
            val lookup = roles?.splitToSequence(',')?.associateBy({ it }, { it }) ?: mapOf()
            return ApiKey(name, key, roles.orEmpty(), lookup)
        }
    }
}
